### PR TITLE
mgr/dashboard: Provide ability to create a user by cloning another

### PR DESCRIPTION
**Pull request by votdev**
auswahl_004

Fixes: https://tracker.ceph.com/issues/25231

Signed-off-by: Volker Theile <vtheile@suse.com>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-list/user-list.component.ts [hunk 1]
```diff
@@ -99,8 +118,8 @@ export class UserListComponent implements OnInit {
     if (sessionUsername === username) {
       this.notificationService.show(
         NotificationType.error,
-        `You are currently authenticated with user "${username}".`,
-        'Cannot Delete User'
+        'Cannot Delete User',
+        `You are currently authenticated with user "${username}".`
```
[the remark this answers is not in the document]
Comment (votdev) [reply src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-list/user-list.component.ts]: Done

File: src/pybind/mgr/dashboard/controllers/user.py [hunk 2]
```diff
@@ -65,6 +65,27 @@ def create(self, username=None, password=None, name=None, email=None, roles=None
...
+            raise DashboardException(msg='Username is required',
+                                     code='username_required',
+                                     component='user')
+        if not new_username:
+            raise DashboardException(msg='New username is required',
+                                     code='new_username_required',
+                                     component='user')
+        try:
+            user = ACCESS_CTRL_DB.clone_user(username, new_username)
+        except UserAlreadyExists:
```
[the remark this answers is not in the document]
Comment (votdev) [reply src/pybind/mgr/dashboard/controllers/user.py]: Done. Check will be done before submission.

**Comment by tspmelo** [comment]
You are missing `:` after `Fixes`, in the commit message.

File: src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-clone-modal/user-clone-modal.component.html [hunk 3]
```diff
@@ -0,0 +1,71 @@
+<div class="modal-header">
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-clone-modal/user-clone-modal.component.html]: Please use the Modal component here.

Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/src/app/core/auth/user-clone-modal/user-clone-modal.component.html]: Just saw we have a few modals that are not using the ModalComponent, so I have created https://tracker.ceph.com/issues/33606.

**Comment by LenzGr** [comment]
Just tested this PR on my local dev environment. Looks good, I agree with @tspmelo that the input focus should be on the new username by default.

Currently, I can only choose a new user name when cloning, which means that full name and email address of the original user account are cloned as well. If these fields have been populated, I'd suggest to either allow to change them in the modal dialogue or leave them empty in the cloned user's information.

**Comment by LenzGr** [comment]
> That does not make sense. If something is cloned, then i want to have an exact copy of the whole object except the field that is required to make it unique (in our case the username).

I'm not sure I agree. The point of being able to clone a user is to minimize the effort of creating a new user from scratch when it comes to roles/privileges in particular. 

However, the full name and email address are user specific usually unique. Why would I want to duplicate these?

**Comment by votdev** [comment]
> The point of being able to clone a user is to minimize the effort of creating a new user from scratch when it comes to roles/privileges in particular.

But what about the password? If we want to modify the full name and email, then we also need to allow the user to modify the password, because this is also unique and tied to a specific user.

If we implement all these things, then the 'Clone' dialog is more or less the same as the 'Create' dialog except the role selection. And really, in the UI this is not an expensive action to click and select the roles for a new user. Finally the 'Clone' feature is obsolete because everything can be done by creating a new user. There is absolutely no benefit of such a new dialog if everything can be done via 'Create' in the same amount of clicks and time.

**Comment by LenzGr** [comment]
> But what about the password? If we want to modify the full name and email, then we also need to allow the user to modify the password, because this is also unique and tied to a specific user.

That's another good point. Cloning a user should also involve either resetting the password or being able to create a new one, otherwise how would you be able to enable that cloned user account to log in?
 
> If we implement all these things, then the 'Clone' dialog is more or less the same as the 'Create' dialog except the role selection. And really, in the UI this is not an expensive action to click and select the roles for a new user. Finally the 'Clone' feature is obsolete because everything can be done by creating a new user. There is absolutely no benefit of such a new dialog if everything can be done via 'Create' in the same amount of clicks and time.

I suspect you are right. As long as the role management does not require a significant amount of effort, there is little value in adding a "clone" function for existing users...

**Comment by votdev** [comment]
This PR is closed because right at the moment the 'Clone' user dialog will look and behave exactly as the 'Create' dialog, so this feature wouldn't improve the workflow somehow.

**Pull request closed without merging**